pkg/controller/endpointslice/utils.go

```diff
@@ -267,6 +270,55 @@ func serviceControllerKey(endpointSlice *discovery.EndpointSlice) (string, error
 	return fmt.Sprintf("%s/%s", endpointSlice.Namespace, serviceName), nil
 }
 
+// setEndpointSliceLabels returns a map with the new endpoint slices labels and true if there was an update
+// The well-known labels IsHeadlessService, LabelServiceName and LabelManagedBy does not generate updates
+// The slice labels must have:
+// 1. the well-known labels discovery.LabelServiceName and discovery.LabelManagedBy
+// 2. the parent service labels
+// 3. the v1.IsHeadlessService label if it is a headless service
+func setEndpointSliceLabels(epSlice *discovery.EndpointSlice, service *corev1.Service) (map[string]string, bool) {
+	updated := false
+	epLabels := make(map[string]string)
+
+	// check if the endpoint slice and the service have the same labels
+	// clone current slice labels but the slice well-known labels
+	for key, value := range epSlice.Labels {
+		if key == discovery.LabelServiceName || key == discovery.LabelManagedBy {
+			continue
+		}
+		// copy endpoint slice labels
+		epLabels[key] = value
+	}
+
+	// if the labels are not identical update the slice with the corresponding service labels
+	if !apiequality.Semantic.DeepEqual(epLabels, service.Labels) {
+		updated = true
+		epLabels = make(map[string]string)
+		// add service labels
+		for key, value := range service.Labels {
+			// don't copy reserved labels
+			if key == discovery.LabelServiceName || key == discovery.LabelManagedBy {
+				klog.Warningf("Service %s/%s using reserved endpoint slices label, skipping label %s: %s", service.Namespace, service.Name, key, value)
+				continue
+			}
+			epLabels[key] = value
+		}
```

Comment:
can we just copy/clone service labels here and rely on lines below to overwrite the reserved keys (`v1.IsHeadlessService`, `discovery.LabelServiceName`, `discovery.LabelManagedBy`)?

Reply:
I think that it should be good to warn about the "bad" behavior, otherwise it will go undetected

Reply:
👍 A warning makes sense here.